src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs

```diff
@@ -0,0 +1,243 @@
+        {
+            foreach (string includePath in difference.RemovedItems)
+            {
+                RemoveFromContextIfPresent(includePath);
+            }
+
+            foreach (string includePath in difference.AddedItems)
+            {
+                AddToContextIfNotPresent(includePath, metadata, isActiveContext);
+            }
+
+            // We Remove then Add changed items to pick up the Linked metadata
+            foreach (string includePath in difference.ChangedItems)
+            {
+                RemoveFromContextIfPresent(includePath);
+                AddToContextIfNotPresent(includePath, metadata, isActiveContext);
+            }
+
+            Assumes.True(difference.RenamedItems.Count == 0, "We should have normalized renames.");
+        }
+
+        private void RemoveFromContextIfPresent(string includePath)
+        {
+            string fullPath = _project.MakeRooted(includePath);
+            if (_paths.Remove(fullPath))
+            {
+                RemoveFromContext(fullPath);
+            }
+        }
+
+        private void AddToContextIfNotPresent(string includePath, IImmutableDictionary<string, IImmutableDictionary<string, string>> metadata, bool isActiveContext)
+        {
+            string fullPath = _project.MakeRooted(includePath);
+            if (_paths.Add(fullPath))
+            {
+                var itemMetadata = metadata.GetValueOrDefault(includePath, ImmutableDictionary<string, string>.Empty);
+                AddToContext(fullPath, itemMetadata, isActiveContext);
+            }
+        }
+
+        private IProjectChangeDiff ResolveDesignTimeConflicts(IComparable designTimeVersion, IProjectChangeDiff designTimeDifference)
+        {
+            DiscardOutOfDateEvaluations(designTimeVersion);
+
+            // Walk all evaluations (if any) that occurred since we launched and resolve the conflicts
+            foreach (VersionedProjectChangeDiff evaluation in _evaluations)
+            {
+                Assumes.True(evaluation.Version.IsLaterThan(designTimeVersion), "Attempted to resolve a conflict between a design-time build and an earlier evaluation.");
+
+                designTimeDifference = ResolveConflicts(evaluation.Difference, designTimeDifference);
+            }
+
+            return designTimeDifference;
+        }
+
+        private IProjectChangeDiff ResolveConflicts(IProjectChangeDiff evaluationDifferences, IProjectChangeDiff designTimeDifferences)
+        {
+            // Remove added items that were removed by later evaluations, and vice versa
+            IImmutableSet<string> added = designTimeDifferences.AddedItems.Except(evaluationDifferences.RemovedItems);
+            IImmutableSet<string> removed = designTimeDifferences.RemovedItems.Except(evaluationDifferences.AddedItems);
+
+            return new ProjectChangeDiff(added, removed, designTimeDifferences.ChangedItems);
```

Comment:
Yeah, I should doc this/assert this - ChangeItems is basically never set during design-time builds. It basically only gets set when you have the same _Include_, but different metadata. Command line arguments themselves are the "include" so if they change slightly then they are considered new items.